File: redact/src/main/java/reuse/redact/Redact.java
package reuse.redact;

import java.util.Objects;
import java.util.function.Function;

public abstract class Redact {
   protected final String pii;

   public Redact(String pii) {
      this.pii = pii;
   }

   @Override
   public String toString() {
      return "<redacted " + redact().apply(pii) + '>';
   }

   protected Function<String, String> redact() {
      return Type.FULL;
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      Redact redact = (Redact) o;
      return pii.equals(redact.pii);
   }

   @Override
   public int hashCode() {
      return Objects.hash(pii);
   }

   public enum Type implements Function<String, String> {
      FULL {
         public String apply(String s) {
            return "***";
         }
      },
      NOOP {
         public String apply(String s) {
            return s;
         }
      },
      HASH {
         public String apply(String s) {
            return Integer.toHexString(s.hashCode());
         }
      },
      LAST_4 {
         public String apply(String s) {
            if (s != null && s.length() > 4) {
               return "*" + s.substring(s.length() - 4);
            } else {
               return s;
            }
         }
      };
   }
}

File: redact-kotlin/src/main/java/us/jsy/redact/Redact.kt
package us.jsy.redact

import java.util.function.Function

/**
 * Helps prevent PII(Personally identifiable information) from being printed or logged by overriding toString().
 */
open class Redact(pii: String) {
    @JvmField
    protected val pii: String = pii

    /**
     * Final to prevent subclasses from accidentally undoing our work
     */
    final override fun toString(): String {
        return "<redacted ${strategy().apply(pii)}>"
    }

    /**
     * Extension point to allow subclasses to redaction strategy. Default to Strategy.FULL
     */
    protected open fun strategy(): Function<String, String> {
        return Strategy.FULL
    }

    // equals and hashCode implementation allow wrapper to be use transparently in place of String
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val redact = other as Redact
        return pii == redact.pii
    }

    override fun hashCode(): Int {
        return pii.hashCode()
    }

}